File: quant1x/exchange/status.go
package exchange

import (
	"os"
)

// getFilenameModifiedTime 获取指定文件的最后修改时间
//
// 参数:
//
//	fname: 文件路径
//
// 返回值:
//
//	*exchange.Timestamp: 文件修改时间的时间戳
//	error: 获取文件信息失败时返回的错误
func getFilenameModifiedTime(fname string) (*Timestamp, error) {
	info, err := os.Lstat(fname)
	if err != nil {
		return nil, err
	}
	tp := NewTimestampFromTime(info.ModTime())
	return &tp, nil
}

// ShouldUpdateFile 检查指定文件是否需要更新
//
// 参数:
//
//	fname - 要检查的文件路径
//
// 返回值:
//
//	bool - 如果文件需要更新则返回true，否则返回false
//
// 注意:
//
//	如果获取文件修改时间失败，默认返回true
func ShouldUpdateFile(fname string) bool {
	modTime, err := getFilenameModifiedTime(fname)
	if err != nil {
		return true
	}
	return CanInitialize(modTime)
}

File: quant1x/exchange/status.py
# -*- coding: UTF-8 -*-
"""
Ported from exchange/status.go

Provides helpers to obtain a file's modification timestamp and
determine whether a file should be updated.

Note: This module will try to call `CanInitialize` from
`quant1x.exchange.session` if available. If that function is
not present, `should_update_file` conservatively returns True.
"""
import os
from datetime import datetime
from typing import Optional

from .timestamp import Timestamp


def get_filename_modified_time(fname: str) -> Timestamp:
    """Get the modification time of a file as a `Timestamp`.

    Raises `OSError` if the file metadata cannot be read.
    """
    info = os.lstat(fname)
    dt = datetime.fromtimestamp(info.st_mtime)
    return Timestamp.from_datetime(dt)


def should_initialize_file(fname: str) -> bool:
    """Check whether the given file should be initialized.

    Mirrors `ShouldInitializeFile` in Go: returns True on metadata error
    and otherwise delegates to `session.CanInitialize`.
    """
    try:
        mod_time = get_filename_modified_time(fname)
    except OSError:
        return True

    from .session import CanInitialize

    return CanInitialize(mod_time)


def should_update_file(fname: str) -> bool:
    """Check whether the given file should be updated.

    Mirrors `ShouldUpdateFile` in Go: on metadata error returns True,
    otherwise uses `session.CheckTradingTimestamp` and returns
    the `update_in_real_time` flag.
    """
    try:
        mod_time = get_filename_modified_time(fname)
    except OSError:
        return True

    from .session import CheckTradingTimestamp

    rs = CheckTradingTimestamp(mod_time)
    return rs.update_in_real_time
